pyrsistent/_checked_types.py

```diff
@@ -45,18 +47,50 @@ def __str__(self):
             missing_fields=', '.join(self.missing_fields))
 
 
+_preserved_iterables = (
+        Enum,
+)
+"""For types that are themselves iterable, we would mistakenly take their
+elements for the type annotation. This set defines types to 'preserve' by not
+iterating over them."""
+
+
+def maybe_type_to_list(t):
+    """Try to coerce a user-supplied type directive into a list of types."""
+    # TODO FIXME: Move type validation inside this function to consolidate it.
+    # Preserve type annotations that are thmeselves iterable.
+    if isinstance(t, type) and issubclass(t, _preserved_iterables):
+        return [t]
+    # Preserve type annotations given as strings.
+    elif isinstance(t, string_types):
+        return [t]
+    else:
+        if not isinstance(t, Iterable):
+            return [t]
+        else:
+            return list(t)
+
+
+def maybe_types_to_list(ts):
+    res = []
+
+    for t in ts:
+        res.extend(maybe_type_to_list(t))
+
+    return tuple(res)
+
+
 def _store_types(dct, bases, destination_name, source_name):
-    def to_list(elem):
-        if not isinstance(elem, Iterable) or isinstance(elem, six.string_types):
-            return [elem]
-        return list(elem)
-
-    dct[destination_name] = to_list(dct[source_name]) if source_name in dct else []
-    dct[destination_name] += sum([to_list(b.__dict__[source_name]) for b in bases if source_name in b.__dict__], [])
-    dct[destination_name] = tuple(dct[destination_name])
-    if not all(isinstance(t, type) or isinstance(t, six.string_types) for t in dct[destination_name]):
+    maybe_types = maybe_types_to_list([
+        d[source_name]
+        for d in ([dct] + [b.__dict__ for b in bases]) if source_name in d
+    ])
+
+    if not all(isinstance(t, type) or isinstance(t, string_types) for t in maybe_types):
```

Comment:
This validation should be moved into `maybe_type_to_list`, but that would require tightening its conditionals or something. Or you could invoke it after all the conditional logic that creates a list of "types." 

Let me know if you have opinions.

Reply:
I guess it could be moved into `maybe_type_to_list`? What do you mean when you say that the conditionals would have to be tightened?

Reply:
Note that, with these changes, all fields depend on `maybe_type_to_list`. `maybe_type_to_list` doesn't raise, so this introduces no regressions.

This assertion was only present for stuff using `_store_types`, which was not all fields. So if we take this validation out and move it into `maybe_types_to_list`, we might see regressions, since `maybe_types_to_list` would be introducing a new assertion for some callers.

If that's acceptable then I'd like to do that, because it further consolidates how we handle user-provided types across all sorts of fields.

I have no idea what I originally meant when I said "the conditionals would have to be tightened" haha, let's just disregard that comment. The existing assertion -- user-provided types must be either `type`s or strings -- seems reasonable to me.

Reply:
I think it should be safe to tighten the check. Even though it might trigger regressions I can't think of any valid code where types should be anything else than type/types or a string.